### ISSUE TITLE
Throttling no returns 'Retry-After' in response header

### ISSUE THREAD

**Issue by ondery** [laravel/framework#34534]
- Laravel Version: 8.5.0
- PHP Version: 7.4
- Database Driver & Version: SQL Server 14

### Description:
When the throttle limit has been reached **with custom response** Laravel no longer returns a 'Retry-After' header element in the response.

**Comment by driesvints** [comment]
Heya, thanks for reporting.

I'll need more info and/or code to debug this further. Please post relevant code like models, jobs, commands, notifications, events, listeners, controller methods, routes, etc. You may use https://paste.laravel.io to post larger snippets or just reply with shorter code snippets. 

Thanks!

**Comment by ondery** [comment]
ok... I tried to retrieve remaining time from the RateLimiter with some other ways but I couldn't... and also because of my custom response **Retry-After** header disappeared.

You can see my RouteServiceProvider.php at below (provider file which is in app\providers folder). On "api" named rate limiter return retry-after info (remaining seconds to retry) in header but "onepermin" named rate limiter which have custom response, doesn't return any retry-after data in header or anywhere else (I couldn't found it).

##################################################################

`<?php

namespace App\Providers;

use Illuminate\Cache\RateLimiting\Limit;

use Illuminate\Foundation\Support\Providers\RouteServiceProvider as ServiceProvider;
use Illuminate\Http\Request;
use Illuminate\Routing\Middleware\ThrottleRequests;
use Illuminate\Support\Facades\RateLimiter;
use Illuminate\Support\Facades\Route;

class RouteServiceProvider extends ServiceProvider
{
    /**
     * The path to the "home" route for your application.
     *
     * This is used by Laravel authentication to redirect users after login.
     *
     * @var string
     */
    public const HOME = '/home';

    /**
     * The controller namespace for the application.
     *
     * When present, controller route declarations will automatically be prefixed with this namespace.
     *
     * @var string|null
     */
     protected $namespace = 'App\\Http\\Controllers';

    /**
     * Define your route model bindings, pattern filters, etc.
     *
     * @return void
     */
    public function boot()
    {
        $this->configureRateLimiting();

        $this->routes(function () {
            Route::prefix('api')
                ->middleware('api')
                ->namespace($this->namespace)
                ->group(base_path('routes/api.php'));

            Route::middleware('web')
                ->namespace($this->namespace)
                ->group(base_path('routes/web.php'));
        });
    }

    /**
     * Configure the rate limiters for the application.
     *
     * @return void
     */
    protected function configureRateLimiting()
    {
        RateLimiter::for('onepermin', function (Request $request) 
            return [
                Limit::perMinute(1)->by($request->input('email'))->response(function ()  use ($request) {
                    $key = $request->fingerprint();
                    $secondsRemaining = RateLimiter::availableIn($key);
                    return response()->json(['errors'=>['message'=> 'Too many request...', 'retryAfter'=> $secondsRemaining]], 429);
                })
            ];
        });


        RateLimiter::for('api', function (Request $request) {
            return Limit::perMinute(60);
        });
    }
}`

#########################################################################################

**Comment by rodrigopedra** [comment]
@ondery 

Try using this:

~~~php

RateLimiter::for('onepermin', function ($request) {
    return Limit::perMinute(1)
        ->by($request->input('email'))
        ->response(function ($request, $headers) {
            return new JsonResponse([
                'errors' => [
                    'message' => 'Too many request...',
                    'retryAfter' => $headers['Retry-After'] ?? 60,
                ],
            ], 429, $headers);
        });
});
~~~

As the Limiter was keyed by `email` this call in your example will return the wrong limit info:

~~~php
$key = $request->fingerprint();
$secondsRemaining = RateLimiter::availableIn($key);
~~~

Also I would remove the `->by($request->input('email'))` from your limiter altogether. Laravel's `RateLimiter` will generate the correct limiter key by each client (browser).

Note in my sample I make use of the `$headers` parameters that are already passed to the callback. Also if you need any `$request` info in the callback you can use the instance passed as the callback's first parameter.

**Comment by ondery** [comment]
I have to limit with an e-mail not by browser; because, if you think that the client clicked the password reset button (to receive an reset email) within a few minutes, first from his desktop and then from his mobile phone (or another device), it will not work as I want....

On the other hand, **$headers** parameter that are already passed to the callback works good. Thank you...

**Comment by rodrigopedra** [comment]
You're welcome =)

**Comment by Stevemoretz** [comment]
> ```
> ->response(function ($request, $headers) {
>             return new JsonResponse([
>                 'errors' => [
>                     'message' => 'Too many request...',
>                     'retryAfter' => $headers['Retry-After'] ?? 60,
>                 ],
>             ], 429, $headers);
>         });
> ```

Why isn't that the default behavior? Why doesn't `retry-after` get sent if it's already in that variable?

**Comment by rodrigopedra** [comment]
@Stevemoretz it is already sent, as a response header. 

If you look at the first code reported, the user was creating a custom response to have the "retry after" also in their response body, but wasn't sending the generated headers on their response.

I am referring to this piece of code:

```php
Limit::perMinute(1)->by($request->input('email'))->response(function ()  use ($request) {
    $key = $request->fingerprint();
    $secondsRemaining = RateLimiter::availableIn($key);
    return response()->json(['errors'=>['message'=> 'Too many request...', 'retryAfter'=> $secondsRemaining]], 429);
})
```

Note they return a new `response()->json()` this response won't have any custom headers automatically added.

In summary:

- Regular rate limiter response does contain the `Retry-After` header with the expected value
- If a developer wants to customize their response, it is up to them to add any needed headers to this custom response

**Comment by Stevemoretz** [comment]
> @Stevemoretz it is already sent, as a response header.
> 
> If you look at the first code reported, the user was creating a custom response to have the "retry after" also in their response body, but wasn't sending the generated headers on their response.
> 
> I am referring to this piece of code:
> 
> ```
> Limit::perMinute(1)->by($request->input('email'))->response(function ()  use ($request) {
>     $key = $request->fingerprint();
>     $secondsRemaining = RateLimiter::availableIn($key);
>     return response()->json(['errors'=>['message'=> 'Too many request...', 'retryAfter'=> $secondsRemaining]], 429);
> })
> ```
> 
> Note they return a new `response()->json()` this response won't have any custom headers automatically added.
> 
> In summary:
> 
> * Regular rate limiter response does contain the `Retry-After` header with the expected value
> * If a developer wants to customize their response, it is up to them to add any needed headers to this custom response

Thanks so much for explaining this! I am using Laravel Lighthouse Graphql and it wasn't returning the header itself, weirdly enough using the snippet you provided add the headers I was wondering what was happening there, it's probably just a bug in lighthouse.

**Comment by rodrigopedra** [comment]
No worries, you can point this thread out to them, if they are missing re-adding the headers on their response Have a nice day =)

**Comment by markosole** [comment]
@rodrigopedra  your suggestion helped a lot and thanks for sharing. Just quick update on my implementation if someone needs or looking for similar.

I am using Laravel 10 and here is how I implemented this.

Add this to your `app\Provides\RouteServiceProvider.php`  in your boot functio which will look like below. I've set limit to 3 attempts per hour. I am using Mobile Device ID as a key and fallback to IP. 

```php

    public function boot(): void
    {

        RateLimiter::for('my-example', function (Request $request) {
            // Device ID (mobile) sent via API endpoint
            $deviceId = $request->device_id ?? $request->ip();
            return Limit::perHour(3)->by($deviceId)
            ->response(function ($request, $headers) {
                return response()->json([
                        'message' => 'You have tried too many times. Try again later.',
                        'status' => 'error',
                        'status_code' => 429,
                        'retryAfter' => $headers['Retry-After'] ?? 60,
                ], 429, $headers);
            });
        });

        // ...
    }

```

## Usage
Define your route in api.php or web.php what ever you use.

```php

Route::post('/your-route', [\App\Http\Controllers\Api\YourController::class, 'test-class'])->middleware(['throttle:my-example']);

```
